pkg/local_object_storage/engine/delete.go

```diff
@@ -58,102 +55,25 @@ func (e *StorageEngine) delete(prm DeletePrm) (DeleteRes, error) {
 		defer elapsed(e.metrics.AddDeleteDuration)()
 	}
 
-	var locked struct {
-		is  bool
-		err apistatus.ObjectLocked
-	}
-	var splitInfo *objectSDK.SplitInfo
-
-	// Removal of a big object is done in multiple stages:
-	// 1. Remove the parent object. If it is locked or already removed, return immediately.
-	// 2. Otherwise, search for all objects with a particular SplitID and delete them too.
-	e.iterateOverSortedShards(prm.addr, func(_ int, sh hashedShard) (stop bool) {
-		var existsPrm shard.ExistsPrm
-		existsPrm.SetAddress(prm.addr)
-
-		resExists, err := sh.Exists(existsPrm)
-		if err != nil {
-			if shard.IsErrRemoved(err) || shard.IsErrObjectExpired(err) {
-				return true
-			}
-
-			var splitErr *objectSDK.SplitInfoError
-			if !errors.As(err, &splitErr) {
-				if !shard.IsErrNotFound(err) {
-					e.reportShardError(sh, "could not check object existence", err)
-				}
-				return false
-			}
-			splitInfo = splitErr.SplitInfo()
-		} else if !resExists.Exists() {
-			return false
-		}
-
-		var shPrm shard.InhumePrm
-		shPrm.MarkAsGarbage(prm.addr)
-		if prm.forceRemoval {
-			shPrm.ForceRemoval()
-		}
-
-		_, err = sh.Inhume(shPrm)
+	if !prm.forceRemoval {
+		locked, err := e.IsLocked(prm.addr)
 		if err != nil {
-			e.reportShardError(sh, "could not inhume object in shard", err)
-
-			locked.is = errors.As(err, &locked.err)
-
-			return locked.is
-		}
-
-		// If a parent object is removed we should set GC mark on each shard.
-		return splitInfo == nil
-	})
-
-	if locked.is {
-		return DeleteRes{}, locked.err
-	}
-
-	if splitInfo != nil {
-		if splitID := splitInfo.SplitID(); splitID != nil {
-			e.deleteChildren(prm.addr, prm.forceRemoval, *splitID)
+			e.log.Warn("deleting an object without full locking check",
+				zap.Error(err),
+				zap.Stringer("addr", prm.addr))
+		} else if locked {
+			var lockedErr apistatus.ObjectLocked
+			return DeleteRes{}, lockedErr
 		}
 	}
 
-	return DeleteRes{}, nil
-}
-
-func (e *StorageEngine) deleteChildren(addr oid.Address, force bool, splitID objectSDK.SplitID) {
-	var fs objectSDK.SearchFilters
-	fs.AddSplitIDFilter(objectSDK.MatchStringEqual, splitID)
-
-	var selectPrm shard.SelectPrm
-	selectPrm.SetFilters(fs)
-	selectPrm.SetContainerID(addr.Container())
-
 	var inhumePrm shard.InhumePrm
-	if force {
+	inhumePrm.MarkAsGarbage(prm.addr)
+	if prm.forceRemoval {
 		inhumePrm.ForceRemoval()
 	}
 
-	e.iterateOverSortedShards(addr, func(_ int, sh hashedShard) (stop bool) {
-		res, err := sh.Select(selectPrm)
-		if err != nil {
-			e.log.Warn("error during searching for object children",
-				zap.Stringer("addr", addr),
-				zap.String("error", err.Error()))
-			return false
-		}
-
-		for _, addr := range res.AddressList() {
-			inhumePrm.MarkAsGarbage(addr)
+	_, err := e.inhumeAddr(prm.addr, inhumePrm)
```

Comment:
Looks like it was a bit more efficient previously with _sorted_ shards.

Reply:
why do you think it is unsorted now?

Reply:
```
        e.iterateOverUnsortedShards(func(sh hashedShard) (stop bool) {
```
in `inhumeAddr`

Reply:
you mean [this](https://github.com/nspcc-dev/neofs-node/blob/0e4f7c919e712ca77e7ce15140599de0e644bb69/pkg/local_object_storage/engine/inhume.go#L150)? it is root object handling, there is no information about link and last objects that can be acquired via root object's ID